resolver/resolver_test.go

```diff
@@ -433,52 +433,145 @@ func TestTruncateEdns0NoSetTruncateBit(t *testing.T) {
-	}
-	msg2 := *truncate(&msg, true, true)
-	if !msg2.Truncated {
-		t.Fatal("Original truncation status was not preseved")
+	before := msg.Len()
+	// test double truncate
+	truncate(msg, max, true)
+	if !msg.Truncated {
+		t.Fatal("Original truncation status was not preserved")
 	}
-	if msg2.Len() != msg.Len() {
+	if msg.Len() != before {
 		t.Fatal("Further modification to already truncated message")
 	}
+	// Add another answer to a truncated message and test that it is then
+	// too large.
 	msg.Answer = append(msg.Answer, genA(1)...)
-	if l := msg.Len(); l < int(maxSize) {
+	if l := msg.Len(); l < int(max) {
 		t.Fatalf("Message to small after adding answers: %d bytes", l)
 	}
 }
 
-func testTruncateNoSetTruncateBit(t *testing.T, maxSize uint16) {
-	orig := *newMessage(maxSize)
-	tmp := orig
-	msg := *truncate(&tmp, true, false)
-	// we expect that the message is currently large
-	// Replace any EDNS0 option that was stripped by the first
-	// truncate. This let's us test "truncate() of a large truncated message".
-	if opt := orig.IsEdns0(); opt != nil {
-		msg.SetEdns0(opt.UDPSize(), opt.Do())
-	}
-	msg2 := *truncate(&msg, true, false)
-	if msg2.Truncated {
-		t.Fatal("Truncate bit not cleared")
+func testTruncateNoSetTruncateBit(t *testing.T, max uint16) {
+	msg := newMessage(max)
+	size := msg.Len()
+	truncate(msg, max, false)
+	if msg.Len() == size {
+		t.Fatal("truncating a large message did not diminish its size")
 	}
-	// expect it to have been truncated
-	if msg2.Len() > int(maxSize) {
+	if msg.Len() > int(max) {
 		t.Fatal("message not truncated")
 	}
+	if msg.Truncated {
+		t.Fatal("Truncate bit set even though setTruncateBit was false")
+	}
+	// We set the Truncate bit on the message and confirm that the bit is
+	// cleared even though the message did not need to be truncated.
+	// This asserts that no matter the message or its size, truncate() will
+	// clear the Truncate bit if setTruncateBit=false is passed to truncate().
+	before := msg.Len()
+	msg.Truncated = true
+	truncate(msg, max, false)
+	if msg.Truncated {
+		t.Fatal("Truncate bit not cleared")
+	}
+	if msg.Len() != before {
+		t.Fatal("message truncated further")
+	}
+	msg.Answer = append(msg.Answer, genA(1)...)
+	if l := msg.Len(); l < int(max) {
+		t.Fatalf("Message to small after adding answers: %d bytes", l)
+	}
+}
+
+func TestTruncateAnswers(t *testing.T) {
+	// Test truncating messages by starting with empty answers and increasing
+	// the number of answers until twice the maximum size is reached and test
+	// that certain invariants are maintained at all times.
+	// Testing boundary conditions only would be typical if the problem space
+	// was large, however in this case since the maximum size is so small
+	// we test the space exhaustively.
+	max := uint16(4096)
+	for nn := 0; ; nn++ {
+		msg := Message(
+			Question("example.com.", dns.TypeA),
+			Header(false, dns.RcodeSuccess),
+			Answers(genA(nn)...))
+		before := msg.Len()
+		if before > int(max)*2 {
```

Comment:
I think we should go up to `int(max)*7)`, reason being, since we're using binary search, we're not really testing the binary search if we end at an even number. This is probably tested enough at `int(max)*7)`, but why not just stick in a few more iterations to be safe :).